### ISSUE TITLE
Warning: Illegal offset type, cannot modify header information

### ISSUE THREAD

**Issue by Tabaluga** [ACSPRI/quexml-docker#1]
Dear @adamzammit,

thanks for your Docker image.
I really appreciate your work and initialize your image with the following `docker-compose.yml` file:

```yml
version: '2'

services:

  quexml:
    image: acspri/quexml
    ports:
      - 8090:80
    volumes:
      - ./IMAGES:/images
```
But queXML can't create the pdf files. The following error appears and I want to ask for help:

```php
Warning: Illegal offset type in isset or empty in /var/lib/tcpdf/include/tcpdf_static.php on line 2457

Warning: Cannot modify header information - headers already sent by (output started at /var/lib/tcpdf/include/tcpdf_static.php:2457) in /var/www/html/index.php on line 83

Warning: Cannot modify header information - headers already sent by (output started at /var/lib/tcpdf/include/tcpdf_static.php:2457) in /var/www/html/index.php on line 84

Warning: Cannot modify header information - headers already sent by (output started at /var/lib/tcpdf/include/tcpdf_static.php:2457) in /var/www/html/index.php on line 85

�_��F���)�V�[��g�K ... (following by lines of garbage sings)
```
best regards

**Comment by adamzammit** [comment]
Does this occur for all queXML files? Did you see if an error also occurs at https://quexmltools.acspri.org.au ?

If it works fine on quexmltools.acspri.org.au - can you please send through the queXML file to me and I will debug it.

Adam

**Comment by Tabaluga** [comment]
Dear @adamzammit,

please excuse my late response.

Yesterday I've had some time to investigate this error. The error occurs only if I want to create an pdf when **I add a queXML style settings file**.

I had no problems with your [online website](https://quexmltools.acspri.org.au/quexml/), creating my questionnaire pdf, even when I add the style settings file here.

Here are both files, as requested:

## questionnaire.xml

```xml
﻿<?xml version="1.0" encoding="UTF-8"?>
<questionnaire xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../quexml.xsd" id="1">
	<title>Lorem Ipsum</title>
	<investigator id="1">
		<name>
			<salutation>lorem ipsum</salutation>
			<firstName>lorem ipsum</firstName>
			<lastName>lorem ipsum</lastName>
		</name>
		<organisation>lorem ipsum</organisation>
		<address>
			<street>lorem ipsum</street>
			<suburb>lorem ipsum</suburb>
			<postcode>88888</postcode>
			<country>lorem ipsum</country>
		</address>
		<phoneNumber>123 456</phoneNumber>
		<emailAddress>lorem@ipsum.com</emailAddress>
		<website>https://lorem.ipsum</website>
	</investigator>
	<dataCollector id="1">
		<name>
			<salutation>lorem ipsum</salutation>
			<firstName>lorem ipsum</firstName>
			<lastName>lorem ipsum</lastName>
		</name>
		<organisation>lorem ipsum</organisation>
		<address>
			<street>lorem ipsum</street>
			<suburb>lorem ipsum</suburb>
			<postcode>88888</postcode>
			<country>lorem ipsum</country>
		</address>
		<phoneNumber>123 456</phoneNumber>
		<emailAddress>lorem@ipsum.com</emailAddress>
		<website>https://lorem.ipsum</website>
	</dataCollector>
	<questionnaireInfo>
		<position>before</position>
		<text>Lorem Ipsum</text>
		<administration>self</administration>
	</questionnaireInfo>
	<section>
		<question hidetitle="true">
			<text>Lorem Ipsum</text>
			<response varName="stud">
				<free>
					<format>text</format>
					<length>24</length>
					<label>Lorem Ipsum</label>
				</free>
			</response>
			<response varName="prfz">
				<free>
					<format>integer</format>
					<length>4</length>
					<label>Lorem Ipsum</label>
				</free>
			</response>
		</question>
	</section>
	<section>
		<question hidetitle="true">
			<text>Lorem Ipsum</text>
			<directive>
				<position>during</position>
				<text>Lorem Ipsum.</text>
				<administration>self</administration>
			</directive>
			<subQuestion varName="afg1">
				<text>Lorem 1</text>
			</subQuestion>
			<subQuestion varName="afg2">
				<text>Lorem 2</text>
			</subQuestion>
			<subQuestion varName="afg3">
				<text>Lorem 3</text>
			</subQuestion>
			<subQuestion varName="afg4">
				<text>Lorem 4</text>
			</subQuestion>
			<subQuestion varName="afg5">
				<text>Lorem 5</text>
			</subQuestion>
			<subQuestion varName="afg6">
				<text>Lorem 6</text>
			</subQuestion>
			<subQuestion varName="afg7">
				<text>Lorem 7</text>
			</subQuestion>
			<subQuestion varName="afg8">
				<text>Lorem 8</text>
			</subQuestion>
			<subQuestion varName="afg9">
				<text>Lorem 9</text>
			</subQuestion>
			<subQuestion varName="afg10">
				<text>Lorem 10</text>
			</subQuestion>
			<subQuestion varName="afg11">
				<text>Lorem 11</text>
			</subQuestion>
			<subQuestion varName="afg12">
				<text>Lorem 12</text>
			</subQuestion>
			<subQuestion varName="afg13">
				<text>Lorem 13</text>
			</subQuestion>
			<subQuestion varName="afg14">
				<text>Lorem 14</text>
			</subQuestion>
			<subQuestion varName="afg15">
				<text>Lorem 15</text>
			</subQuestion>
			<subQuestion varName="afg16">
				<text>Lorem 16</text>
			</subQuestion>
			<subQuestion varName="afg17">
				<text>Lorem 17</text>
			</subQuestion>
			<subQuestion varName="afg18">
				<text>Lorem 18</text>
			</subQuestion>
			<subQuestion varName="afg19">
				<text>Lorem 19</text>
			</subQuestion>
			<subQuestion varName="afg20">
				<text>Lorem 20</text>
			</subQuestion>
			<subQuestion varName="afg21">
				<text>Lorem 21</text>
			</subQuestion>
			<subQuestion varName="afg22">
				<text>Lorem 22</text>
			</subQuestion>
			<subQuestion varName="afg23">
				<text>Lorem 23</text>
			</subQuestion>
			<subQuestion varName="afg24">
				<text>Lorem 24</text>
			</subQuestion>
			<response varName="pkt" split="1">
				<fixed>
					<category>
						<label>A</label>
						<value>1</value>
					</category>
					<category>
						<label>B</label>
						<value>2</value>
					</category>
					<category>
						<label>C</label>
						<value>3</value>
					</category>
					<category>
						<label>D</label>
						<value>4</value>
					</category>
					<category>
						<label>E</label>
						<value>5</value>
					</category>
				</fixed>
			</response>
		</question>
	</section>
</questionnaire>
```

## questionnaire_style.xml

```xml
<?xml version="1.0"?>
<queXMLPDFStyle>
  <AllowSplittingSingleChoiceVertical>false</AllowSplittingSingleChoiceVertical>
  <AllowSplittingSingleChoiceHorizontal>false</AllowSplittingSingleChoiceHorizontal>
  <QuestionTextRightMargin>40</QuestionTextRightMargin>
  <Columns>1</Columns>
  <QuestionnaireInfoMargin>5</QuestionnaireInfoMargin>
  <SingleResponseHorizontalHeight>8</SingleResponseHorizontalHeight>
  <SingleResponseAreaHeight>9</SingleResponseAreaHeight>
  <BackgroundColourQuestion>255</BackgroundColourQuestion>
  <BackgroundColourSection>255</BackgroundColourSection>
  <AllowSplittingVas>false</AllowSplittingVas>
  <AllowSplittingMatrixText>false</AllowSplittingMatrixText>
  <DisplayCodeValues>false</DisplayCodeValues>
  <AllowSplittingResponses>false</AllowSplittingResponses>
  <SectionHeight>18</SectionHeight>
  <ResponseLabelFontSize>7.5,6.5</ResponseLabelFontSize>
  <ResponseLabelFontSizeSmall>6.5</ResponseLabelFontSizeSmall>
  <ResponseTextFontSize>10</ResponseTextFontSize>
  <CodeValuesFontSize>8</CodeValuesFontSize>
  <EmbedFonts>true</EmbedFonts>
  <Style>    td.questionTitle {font-weight:bold; font-size:12pt;}&#xD;
    td.questionTitleSkipTo {font-weight:bold; font-size:16pt;}&#xD;
    td.questionText {font-weight:bold; font-size:12pt;} &#xD;
    td.questionSpecifier {font-weight:normal; font-size:12pt;} &#xD;
    td.vasLabelLeft {font-weight:bold; font-size:10pt; text-align:center;}&#xD;
    td.vasLabelRight {font-weight:bold; font-size:10pt; text-align:center;}&#xD;
    td.questionHelp {font-weight:normal; text-align:right; font-style:italic; font-size:8pt;}&#xD;
    td.questionHelpAfter {text-align:center; font-weight:bold; font-size:10pt;}&#xD;
    td.questionHelpBefore {text-align:center; font-weight:bold; font-size:12pt;}&#xD;
    td.responseAboveText {font-weight:normal; font-style:normal; text-align:left; font-size:12pt;} &#xD;
    td.matrixResponseGroupLabel {font-weight:normal; font-style:normal; text-align:left; font-size:12pt;}&#xD;
    span.sectionTitle {font-size:18pt; font-weight:bold;} &#xD;
    span.questionnaireTitle {font-size:22pt; font-weight:bold; text-align:center;} &#xD;
    span.sectionDescription {font-size:14pt; font-weight:bold;} &#xD;
    div.sectionInfo {font-style:normal; font-size:10pt; text-align:left; font-weight:normal;}&#xD;
    td.questionnaireInfo {font-size:14pt; text-align:center; font-weight:bold;}&#xD;
    </Style>
  <EdgeDetectionFormat>lines</EdgeDetectionFormat>
  <PageFormat>A4</PageFormat>
  <PageOrientation>P</PageOrientation>
</queXMLPDFStyle>
```

Thanks for your support and your help.
best regards

**Comment by Tabaluga** [comment]
Dear @adamzammit,

I've reached another *workaround* to complete the generation of the pdf file with *a queXML style settings file added*, based upon my investigation - but it's not a solution.

The following errors:

```php
Warning: Illegal offset type in isset or empty in /var/lib/tcpdf/include/tcpdf_static.php on line 2457

Warning: Cannot modify header information - headers already sent by (output started at /var/lib/tcpdf/include/tcpdf_static.php:2457)
...
```

occurs while TCPDF want to get the page format ([Source](https://github.com/tecnickcom/TCPDF/blob/master/include/tcpdf_static.php)):

```php
...
/**
 * Get page dimensions from format name.
 * @param $format (mixed) The format name @see self::$page_format<ul>
 * @return array containing page width and height in points
 * @since 5.0.010 (2010-05-17)
 * @public static
 */
public static function getPageSizeFromFormat($format) {
	if (isset(self::$page_formats[$format])) {
		return self::$page_formats[$format];
	}
	return self::$page_formats['A4'];
}
...
```
My only idea was - since I the default value is the same as mine - to delete the `<PageFormat>` setting and try it again, without to modify header information:

## questionnaire_style_updated.xml

```xml
<?xml version="1.0"?>
<queXMLPDFStyle>
  <AllowSplittingSingleChoiceVertical>false</AllowSplittingSingleChoiceVertical>
  <AllowSplittingSingleChoiceHorizontal>false</AllowSplittingSingleChoiceHorizontal>
  <QuestionTextRightMargin>40</QuestionTextRightMargin>
  <Columns>1</Columns>
  <QuestionnaireInfoMargin>5</QuestionnaireInfoMargin>
  <SingleResponseHorizontalHeight>8</SingleResponseHorizontalHeight>
  <SingleResponseAreaHeight>9</SingleResponseAreaHeight>
  <BackgroundColourQuestion>255</BackgroundColourQuestion>
  <BackgroundColourSection>255</BackgroundColourSection>
  <AllowSplittingVas>false</AllowSplittingVas>
  <AllowSplittingMatrixText>false</AllowSplittingMatrixText>
  <DisplayCodeValues>false</DisplayCodeValues>
  <AllowSplittingResponses>false</AllowSplittingResponses>
  <SectionHeight>18</SectionHeight>
  <ResponseLabelFontSize>7.5,6.5</ResponseLabelFontSize>
  <ResponseLabelFontSizeSmall>6.5</ResponseLabelFontSizeSmall>
  <ResponseTextFontSize>10</ResponseTextFontSize>
  <CodeValuesFontSize>8</CodeValuesFontSize>
  <EmbedFonts>true</EmbedFonts>
  <Style>    td.questionTitle {font-weight:bold; font-size:12pt;}&#xD;
    td.questionTitleSkipTo {font-weight:bold; font-size:16pt;}&#xD;
    td.questionText {font-weight:bold; font-size:12pt;} &#xD;
    td.questionSpecifier {font-weight:normal; font-size:12pt;} &#xD;
    td.vasLabelLeft {font-weight:bold; font-size:10pt; text-align:center;}&#xD;
    td.vasLabelRight {font-weight:bold; font-size:10pt; text-align:center;}&#xD;
    td.questionHelp {font-weight:normal; text-align:right; font-style:italic; font-size:8pt;}&#xD;
    td.questionHelpAfter {text-align:center; font-weight:bold; font-size:10pt;}&#xD;
    td.questionHelpBefore {text-align:center; font-weight:bold; font-size:12pt;}&#xD;
    td.responseAboveText {font-weight:normal; font-style:normal; text-align:left; font-size:12pt;} &#xD;
    td.matrixResponseGroupLabel {font-weight:normal; font-style:normal; text-align:left; font-size:12pt;}&#xD;
    span.sectionTitle {font-size:18pt; font-weight:bold;} &#xD;
    span.questionnaireTitle {font-size:22pt; font-weight:bold; text-align:center;} &#xD;
    span.sectionDescription {font-size:14pt; font-weight:bold;} &#xD;
    div.sectionInfo {font-style:normal; font-size:10pt; text-align:left; font-weight:normal;}&#xD;
    td.questionnaireInfo {font-size:14pt; text-align:center; font-weight:bold;}&#xD;
    </Style>
  <EdgeDetectionFormat>lines</EdgeDetectionFormat>
  <PageOrientation>P</PageOrientation>
</queXMLPDFStyle>
```

After this change, the generation works but it's not the right solution for me.

Thanks for your support and your help.
best regards

**Comment by adamzammit** [comment]
Thanks for this debugging info.

I have made some changes to the queXML code and released a new docker image that I have tagged as acspri/quexml:1.7.0

Can you please try with this image and let me know how you go?

Please note that by default an existing image will not be overridden so please try a "fresh" docker container.

**Comment by Tabaluga** [comment]
Dear @adamzammit,

thanks for your effort.

I've deleted all queXML containers and images:

```sh
docker stop quexml
docker rm quexml
docker rmi quexml
```

and initialize your new image with the following `docker-compose.yml` file:

```yml
version: '2'

services:

  quexml:
    image: acspri/quexml:1.7.0
    ports:
      - 8090:80
    volumes:
      - ./IMAGES:/images
```

After this I've tried to generate the pdf with the *queXML style settings file added* but the following errors appear:

```php
Warning: Illegal offset type in isset or empty in /var/lib/tcpdf/include/tcpdf_static.php on line 2457

Fatal error: Uncaught Error: Call to undefined function gT() in /var/www/html/quexmlpdf.php:2098 Stack trace: #0 /var/www/html/index.php(76): queXMLPDF->createqueXML('<?xml version="...') #1 {main} thrown in /var/www/html/quexmlpdf.php on line 2098
```

Thanks for your support and your help.
best regards

**Comment by adamzammit** [comment]
Apologies there was a bug in the 1.7.0 release.

I have now released 1.7.1 - please try with:

acspri/quexml:1.7.1

Adam

**Comment by Tabaluga** [comment]
Dear @adamzammit,
please excuse my late response.
Everything works fine without errors.
Thanks for your support and your help. 👍 
best regards
